Feature/508 따닥 방지 구현 (Redis 상태 확인 후 중복 클릭 방지 인터셉터 등록)

higukang opened the pull request:
# 🚀 Pull Request

**[Redis 상태 확인 후 중복 클릭 방지 인터셉터 등록]**

## #️⃣ 연관된 이슈

#508

## 📋 작업 내용
  - Redis 상태 확인 후 중복 클릭 방지 인터셉터 등록
  - `InterceptorMvcConfig`에서 `redisConnectionCheckService.isRedisAvailable()`을 활용하여 Redis가 활성화된 경우에만 `DuplicatedClickPrevent` 인터셉터 등록
  - 모든 요청(`/**`)에 대해 중복 클릭 방지를 적용하되, 로그인(`/login`), 토큰 재발급(`/reissue`), API 문서(`/swagger-ui/**`, `/v3/api-docs/**`) 요청 및 GET요청은 제외
  - 불필요한 인터셉터 등록을 방지하여 성능 최적화 및 안정성 강화

## ✅ 변경 사항
- `InterceptorMvcConfig`에 Redis 상태 확인 로직 추가
- `DuplicatedClickPrevent` 인터셉터를 특정 조건에서만 적용
- `excludePathPatterns()`을 통해 특정 엔드포인트 예외 처리

## 🔍 테스트 방법
1. Redis가 활성화된 상태에서 `/login`을 여러 번 요청 → 중복 요청이 차단되지 않아야 함
2. Redis가 활성화된 상태에서 다른 API(회원정보 수정 등) 요청을 빠르게 여러 번 수행 → 429 응답 확인
3. Redis가 비활성화된 상태에서 서버 실행 → 인터셉터가 등록되지 않아야 함

## 🏷 기타 사항
- 중복 요청 방지 로직이 정상적으로 동작하는지 추가적인 테스트 필요
- 인터셉터 적용 범위 조정 가능성 검토

## 설명

```java
public boolean checkAndSetRequest(String memberId, String requestURI, String method) {
    // 요청별로 고유한 키 생성: "CLICK:MEMBERID:요청유형:요청경로:MEMBERID"
    String key = "CLICK:" + memberId + ":" + method + ":" + requestURI + ":" + memberId;
    Boolean success = stringRedisTemplate.opsForValue()
            .setIfAbsent(key, "LOCK", 1, TimeUnit.SECONDS);

    return success != null && success;
}

// setIfAbsent(key, "LOCK", 1, TimeUnit.SECONDS); 부분은 Redis에 Lock을 설정하는 코드
// setIfAbsent(Redis SETNX) 명령어를 사용하여 중복 요청을 방지 : 같은 키가 존재하지 않을 경우에만 값을 설정하며, 1초 동안 유지되도록 TTL(Time-To-Live)을 설정
// 동일한 요청이 1초 내에 다시 들어올 경우 중복 요청으로 간주하여 차단합니다.

higukang committed:
feat: 중복 요청 방지를 위한 429 Too Many Requests 오류 코드 추가

- ErrorCode.java에 ALREADY_PROCESSING_REQUEST(429 Too Many Requests) 추가
- 중복 요청이 감지될 경우 클라이언트가 적절한 처리(재시도 지연 등)를 할 수 있도록 수정

#Issue: 508

higukang committed:
feat: Redis 중복 클릭 방지 저장소 추가 및 초기화 개선

- 중복 클릭 방지를 위한 Redis 저장소(2번 DB) 추가
- `redisConnectionFactoryForDuplicatedClick()` 추가 및 `afterPropertiesSet()` 호출로 초기화 보장
- `@Primary`를 `redisTemplateForInactiveMembers()`에 적용하여 기본 RedisTemplate 설정 명확화
- Redis 설정 개선으로 중복 요청 방지 기능 강화

#Issue: 508

higukang committed:
feat: 중복 요청 예외 처리 추가

- `DuplicateRequestException` 클래스 생성하여 중복 요청 처리 전용 예외 정의
- 기본적으로 `ErrorCode.ALREADY_PROCESSING_REQUEST`를 사용하도록 설정
- 커스텀 메시지를 지원하는 생성자 추가
- 중복 요청 발생 시 명확한 예외 처리 가능하도록 개선

#Issue: 508

higukang committed:
feat: JWTFilter에서 memberId 요청 속성 추가

- `JWTFilter`에서 `request.setAttribute("memberId", id);` 추가하여 요청 속성에 사용자 ID 설정
- 이후 핸들러 및 인터셉터에서 `memberId`를 활용할 수 있도록 개선
- 보안 및 인증 관련 로직의 일관성을 유지하면서 중복 요청 방지 및 로깅 등에 활용 가능

#Issue: 508

higukang committed:
refactor: MemberServiceImpl 트랜잭션 관리 최적화

- `@Transactional(readOnly = true)`을 조회 메서드(`getMemberById`, `getAllMembersWithFilters`, `getMemberStatus`)에 적용하여 성능 최적화
- `@Transactional`을 데이터 변경 메서드(`registerMember`, `updateMember`, `changePassword`, `deleteMember`, `activateMember`, `deactivateMember`)에 명시적으로 적용하여 트랜잭션 관리 강화
- `redisTemplateForInactiveMemers` 오타 수정 (`redisTemplateForInactiveMembers`
- 전체적인 코드 정리 및 유지보수성 개선

#Issue: 508

higukang committed:
feat: 중복 요청 방지를 위한 인터셉터 `DuplicatedClickPrevent` 추가

- `HandlerInterceptor`를 구현하여 중복 클릭 방지 기능 추가
- Redis를 활용하여 `setIfAbsent()` 방식으로 중복 요청 차단 (TTL 3초)
- GET 요청은 예외적으로 통과하도록 처리
- `memberId`를 기반으로 요청 식별 및 중복 체크 수행
- 중복 요청 발생 시 `DuplicateRequestException` 예외 발생

#Issue: 508

higukang committed:
feat: Redis 상태 확인 후 중복 클릭 방지 인터셉터 등록

- `InterceptorMvcConfig`에서 `redisConnectionCheckService.isRedisAvailable()`을 활용하여 Redis가 활성화된 경우에만 `DuplicatedClickPrevent` 인터셉터 등록
- 모든 요청(`/**`)에 대해 중복 클릭 방지를 적용하되, 로그인(`/login`), 토큰 재발급(`/reissue`), API 문서(`/swagger-ui/**`, `/v3/api-docs/**`) 요청은 제외
- 불필요한 인터셉터 등록을 방지하여 성능 최적화 및 안정성 강화

#Issue: 508

higukang committed:
feat: 특정 경로의 PUT 요청 중복 클릭 방지 제외

- /projects/{id}/progress-steps/orders 경로의 PUT 요청을 중복 클릭 방지 예외 처리
- preHandle 메서드에서 해당 경로의 요청을 필터링하여 바로 통과하도록 수정
- Redis의 중복 클릭 방지 TTL을 3초에서 1초로 조정

#Issue: 508

oowtl commented on module-api/src/main/java/com/checkping/api/auth/interceptor/DuplicatedClickPrevent.java:
```diff
@@ -0,0 +1,64 @@
+
+    public boolean checkAndSetRequest(String email) {
+        String key = PREFIX + email;
+        Boolean success = stringRedisTemplate.opsForValue()
```
Lock 을 거는 시간에 대한 코드인가요?
PR 설명에 추가 부탁드려도 될까요?

higukang replied:
넵 맞습니다!
추가 설명 PR 에 추가해두겠습니다

oowtl reviewed (approved):
레디스를 활용하여 중복 요청을 방지하는 기능을 확인하였습니다.
코드에서 중복 방지를 하는 시간을 어디에 셋팅을 하는지 주석에 추가되었으면 좋겠습니다.

higukang committed:
chore 중복 클릭 방지 로직 개선: 요청 유형 및 요청 경로 포함

- Redis 키 구조 변경: CLICK:{memberId}:{요청유형}:{요청경로}:{memberId} 형식 적용
- 중복 체크 시 요청 유형(HTTP Method)도 고려하여 중복 방지 로직 강화
- 특정 PUT 요청(`/projects/{id}/progress-steps/orders`)은 예외 처리 유지
- 이메일 기반 키에서 `memberId` 기반으로 변경할 떄 수정되지 않은 변수명 수정

#Issue: 508
